guacamole/src/main/java/org/apache/guacamole/rest/connection/ConnectionResource.java

```diff
@@ -150,18 +153,26 @@ public ConnectionResource(@Assisted UserContext userContext,
      * @throws GuacamoleException
      *     If an error occurs while retrieving the connection history.
      */
-    @GET
+    @SuppressWarnings("deprecation")
     @Path("history")
-    public List<APIConnectionRecord> getConnectionHistory()
+    public ConnectionHistoryResource getConnectionHistory()
             throws GuacamoleException {
 
-        // Retrieve the requested connection's history
-        List<APIConnectionRecord> apiRecords = new ArrayList<APIConnectionRecord>();
-        for (ConnectionRecord record : connection.getHistory())
-            apiRecords.add(new APIConnectionRecord(record));
-
-        // Return the converted history
-        return apiRecords;
+        try {
+            return new ConnectionHistoryResource(connection.getConnectionHistory());
+        }
+        catch (GuacamoleUnsupportedException e) {
+            try {
+                List<ConnectionRecord> history = Collections.emptyList();
+                for (ConnectionRecord record : connection.getHistory()) {
+                    history.add(record);
+                }
+                return new ConnectionHistoryResource(new SimpleActivityRecordSet<>(history));
+            }
+            catch (GuacamoleUnsupportedException ex) {
+                return new ConnectionHistoryResource(new SimpleActivityRecordSet<>());
+            }
+        }
```

Comment:
A couple of points here:

1. The logic here is involved enough that I think it deserves some comments.
2. It's worth logging these exceptions with context at the debug level.
3. Because each `try` would `return` from the function entirely, you can avoid nesting the `try`/`catch` deeper and deeper with something like:

   ```java
   // Try to do it
   try {
      return doSomething();
   }
   catch (GuacamoleUnsupportedException e) {
       logger.debug("Useful context.", e);
   }

   // Well that didn't work
   try {
      return doSomethingElse();
   }
   catch (GuacamoleUnsupportedException e) {
       logger.debug("Useful context.", e);
   }

   // Darn
   return doLoseAllHope();
   ```